File: python/0141.py
from data_strcutres import ListNode


class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        #  return self.use_set(head)
        return self.tortoise_and_hare(head)

    # O(1) space complexity
    def tortoise_and_hare(self, head: ListNode) -> bool:
        try:
            slow = head.next
            fast = head.next.next
            while slow != fast:
                slow = slow.next
                fast = fast.next.next
            return True
        except:
            return False


    def use_set(self, head: ListNode) -> bool:
        visited = set()
        while head:
            if head in visited:
                return True
            else:
                visited.add(head)
                head = head.next
        return False

File: cpp/0141.cpp
struct ListNode {
    int val;
    ListNode* next;
    ListNode(int v) : val(v) {}
};

class Solution {
public:
    bool hasCycle(ListNode* head) {
        if (!head) return false;
        ListNode* slow = head;
        ListNode* fast = head;
        while(fast->next && fast->next->next) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) {
                return true;
            }
        }
        return false;
    }
};
